ecopod/sendRawTransactionConditional.md

```diff
@@ -0,0 +1,53 @@
+# Purpose
+
+To fully unlock ERC-4337 on the OP-Stack.
+
+# Summary
+
+By providing an auxilliary transaction submission mechanism, [eth_sendRawTransactionConditional](https://notes.ethereum.org/@yoav/SkaX2lS9j), we can enable Bundlers to submit Entrypoint transactions with stronger guarantees, avoiding costly inadvertent reverts.
+
+We can implement this endpoint in op-geth with an additional layer of external validation to ensure this endpoint is safe from DoS attacks, does not supersede `eth_sendRawTransaction`, and can be deprecated if native account abstraction efforts in the future absolves the need for this endpoint.
+
+**Note:** As a result of [EIP-3074](https://eips.ethereum.org/EIPS/eip-3074), the interop project is also exploring external validation as a mechanism to preventing DoS attack. A low-stakes scoped version of this for `eth_sendRawTransactionConditional` is a good opportunity for learning.
+
+
+# Problem Statement + Context
+
+Bundlers aggregate UserOps from an external mempool into a single transaction sent to the enshrined 4337 `Entrypoint` contract. It is the Bundler's responsibility to ensure every UserOp in the transaction will not revert when its validation step is executed. In between tx submission and inclusion, there is a chance network state changes causes the transaction to revert, even after local bundler validation.
+
+With private Bundler mempools, the likelihood of these reverts are small. However shared 4337 mempools are coming into [production](https://medium.com/etherspot/decentralized-future-erc-4337-shared-mempool-launches-on-ethereum-b6c860072f41), launched on Ethereum and some L2 testnets, decentralizing 4337 infrastructure. This is possible with no change on Ethereum through special block builders like Flashbots which support atomic transaction inclusion. An endpoint like `eth_sendRawTransactionConditional` is required to launch a shared mempools on L2 as there's only a single block builder, the sequencer, and the likelihood of reverts with a shared mempool would be too high of a cost for bundlers to operate.
```

Comment:
I think its important to be explicit around the problem of "multiple bundlers include the same user op". This is the precise reason why the public 4337 mempool needs a situation like this

Reply:
Flashbots does not solve this problem

Reply:
+1, I'd like to understand this better too

Reply:
sounds good, i'll elaborate on this. And I believe flashbots does solve this problem through `eth_sendBundle`? Multiple bundlers can submit txs to flashbots which are omitted if they revert

Reply:
`eth_sendBundle` has the ability to express "do not include my tx if it reverts" (or it may just be implicit via that endpoint) but my current understanding is that the top level 4337 bundle will not revert if there is a single reverting userop. that particular userop will just revert and the others will be fine. This would be good to dig into to double check but seems like a huge dos vector if a single userop could cause the entire bundle to revert. If the top level tx (4337 bundle) doesn't revert but only its subcalls revert (userops) then `eth_sendBundle` does not solve the problem

Reply:
> my current understanding is that the top level 4337 bundle will not revert if there is a single reverting userop.

It depends what part of the UserOp is reverting. If the validation phase of the UserOp reverts, the entire transaction does reverts, even if only a single UserOp fails

The execution phase of the UserOp is allowed to revert without concern to the top level tx

Reply:
Got it, thanks! Could you add this to the doc so other people don't have to ask the same q?